File: blooming_directed_graph_agents/trigger_process_agent/python/agent.py
"""
blooming_directed_graph_trigger_process_agent – Python.
IEEE 752 64-bit geometrica Z field (zed): see docs/IEEE_752_64BIT_GEOMETRICA_Z_FIELD_ZED_SPECIFICATIONS.md
Triggers processes across languages, projects, repositories, clusters, models.
Action types: workflow_dispatch, job_submit, git_push, git_fetch, fetch_merge, push_merge, pull_merge, run_script, notify.
"""

import os
import json
import subprocess
import urllib.request
import urllib.error
from typing import Dict, Any, Optional


GITHUB_API = "https://api.github.com"


def trigger(node: Dict[str, Any], action: str, payload: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    payload = payload or {}
    if action == "workflow_dispatch":
        return _trigger_workflow_dispatch(node, payload)
    if action == "git_push":
        return _trigger_git(node, "push", payload)
    if action == "git_fetch":
        return _trigger_git(node, "fetch", payload)
    if action == "fetch_merge":
        return _trigger_fetch_merge(node, payload)
    if action == "push_merge":
        return _trigger_push_merge(node, payload)
    if action == "pull_merge":
        return _trigger_git(node, "pull", payload)
    if action == "job_submit":
        return _trigger_hpcc_job(node, payload)
    if action == "run_script":
        return _trigger_run_script(node, payload)
    if action == "notify":
        return {"ok": True, "message": "notify not implemented (platform-specific)"}
    return {"ok": False, "error": f"unknown action: {action}"}


def _trigger_workflow_dispatch(node: Dict[str, Any], payload: Dict[str, Any]) -> Dict[str, Any]:
    owner = node.get("owner") or payload.get("owner")
    repo = node.get("repo") or payload.get("repo") or node.get("slug")
    workflow_id = payload.get("workflow_id")
    ref = payload.get("ref", "main")
    if not owner or not repo or not workflow_id:
        return {"ok": False, "error": "missing owner/repo/workflow_id"}
    token = os.environ.get("GITHUB_TOKEN")
    if not token:
        return {"ok": False, "error": "GITHUB_TOKEN not set"}
    url = f"{GITHUB_API}/repos/{owner}/{repo}/actions/workflows/{workflow_id}/dispatches"
    body = {"ref": ref}
    if payload.get("inputs"):
        body["inputs"] = payload["inputs"]
    req = urllib.request.Request(url, data=json.dumps(body).encode(), method="POST")
    req.add_header("Accept", "application/vnd.github.v3+json")
    req.add_header("Authorization", f"Bearer {token}")
    req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            return {"ok": resp.status < 400}
    except urllib.error.HTTPError as e:
        return {"ok": False, "error": f"HTTP {e.code}"}
    except Exception as e:
        return {"ok": False, "error": str(e)}


def _trigger_git(node: Dict[str, Any], cmd: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    path = node.get("path") or payload.get("path")
    if not path or not os.path.isdir(path):
        return {"ok": False, "error": "missing path"}
    args = payload.get("args") or []
    r = subprocess.run(["git", "-C", path, cmd] + list(args), capture_output=True)
    return {"ok": r.returncode == 0}


def _trigger_fetch_merge(node: Dict[str, Any], payload: Dict[str, Any]) -> Dict[str, Any]:
    path = node.get("path") or payload.get("path")
    if not path or not os.path.isdir(path):
        return {"ok": False, "error": "missing path"}
    remote = payload.get("remote", "origin")
    branch = payload.get("branch", "main")
    r1 = subprocess.run(["git", "-C", path, "fetch", remote], capture_output=True)
    if r1.returncode != 0:
        return {"ok": False}
    r2 = subprocess.run(["git", "-C", path, "merge", f"{remote}/{branch}"], capture_output=True)
    return {"ok": r2.returncode == 0}


def _trigger_push_merge(node: Dict[str, Any], payload: Dict[str, Any]) -> Dict[str, Any]:
    path = node.get("path") or payload.get("path")
    if not path or not os.path.isdir(path):
        return {"ok": False, "error": "missing path"}
    remote = payload.get("remote", "origin")
    branch = payload.get("branch", "main")
    subprocess.run(["git", "-C", path, "merge", branch], capture_output=True)
    args = payload.get("args") or []
    r = subprocess.run(["git", "-C", path, "push", remote] + list(args), capture_output=True)
    return {"ok": r.returncode == 0}


def _trigger_hpcc_job(node: Dict[str, Any], payload: Dict[str, Any]) -> Dict[str, Any]:
    path = node.get("path") or payload.get("path")
    script = payload.get("script", "job.sh")
    if not path:
        return {"ok": False, "error": "missing path"}
    r = subprocess.run(["sbatch", os.path.join(path, script)], cwd=path, capture_output=True)
    return {"ok": r.returncode == 0}


def _trigger_run_script(node: Dict[str, Any], payload: Dict[str, Any]) -> Dict[str, Any]:
    path = node.get("path") or payload.get("path")
    script = payload.get("script")
    if not path or not script:
        return {"ok": False, "error": "missing path/script"}
    r = subprocess.run(script, shell=True, cwd=path)
    return {"ok": r.returncode == 0}

File: blooming_directed_graph_agents/trigger_process_agent/typescript/agent.ts
/**
 * blooming_directed_graph_trigger_process_agent – TypeScript. Namespace: LangFlow.
 * IEEE 752 64-bit geometrica Z field (zed): see docs/IEEE_752_64BIT_GEOMETRICA_Z_FIELD_ZED_SPECIFICATIONS.md
 */

import { execSync } from "child_process";
import * as fs from "fs";
import * as path from "path";

const GITHUB_API = "https://api.github.com";

export interface Node {
  name: string;
  path: string;
  owner: string | null;
  repo: string | null;
  slug: string;
}

export interface TriggerResult {
  ok: boolean;
  error?: string;
  message?: string;
}

export async function trigger(
  node: Node,
  action: string,
  payload: Record<string, unknown> = {}
): Promise<TriggerResult> {
  switch (action) {
    case "workflow_dispatch":
      return await triggerWorkflowDispatch(node, payload);
    case "git_push":
      return triggerGit(node, "push", payload);
    case "git_fetch":
      return triggerGit(node, "fetch", payload);
    case "fetch_merge":
      return triggerFetchMerge(node, payload);
    case "push_merge":
      return triggerPushMerge(node, payload);
    case "pull_merge":
      return triggerGit(node, "pull", payload);
    case "job_submit":
      return triggerHpccJob(node, payload);
    case "run_script":
      return triggerRunScript(node, payload);
    case "notify":
      return { ok: true, message: "notify not implemented (platform-specific)" };
    default:
      return { ok: false, error: `unknown action: ${action}` };
  }
}

async function triggerWorkflowDispatch(
  node: Node,
  payload: Record<string, unknown>
): Promise<TriggerResult> {
  const owner = (node.owner || payload.owner) as string;
  const repo = (node.repo || payload.repo || node.slug) as string;
  const workflowId = payload.workflow_id as string;
  const ref = (payload.ref as string) || "main";
  if (!owner || !repo || !workflowId)
    return { ok: false, error: "missing owner/repo/workflow_id" };
  const token = process.env.GITHUB_TOKEN;
  if (!token) return { ok: false, error: "GITHUB_TOKEN not set" };
  const url = `${GITHUB_API}/repos/${owner}/${repo}/actions/workflows/${workflowId}/dispatches`;
  const body: Record<string, unknown> = { ref };
  if (payload.inputs) body.inputs = payload.inputs;
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: {
        Accept: "application/vnd.github.v3+json",
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    return { ok: res.ok };
  } catch (e) {
    return { ok: false, error: String(e) };
  }
}

function triggerGit(
  node: Node,
  cmd: string,
  payload: Record<string, unknown>
): TriggerResult {
  const p = (node.path || payload.path) as string;
  if (!p || !fs.existsSync(p)) return { ok: false, error: "missing path" };
  const args = (payload.args as string[]) || [];
  try {
    execSync(`git -C ${JSON.stringify(p)} ${cmd} ${args.join(" ")}`, { stdio: "inherit" });
    return { ok: true };
  } catch {
    return { ok: false };
  }
}

function triggerFetchMerge(node: Node, payload: Record<string, unknown>): TriggerResult {
  const p = (node.path || payload.path) as string;
  if (!p || !fs.existsSync(p)) return { ok: false, error: "missing path" };
  const remote = (payload.remote as string) || "origin";
  const branch = (payload.branch as string) || "main";
  try {
    execSync(`git -C ${JSON.stringify(p)} fetch ${remote}`, { stdio: "inherit" });
    execSync(`git -C ${JSON.stringify(p)} merge ${remote}/${branch}`, { stdio: "inherit" });
    return { ok: true };
  } catch {
    return { ok: false };
  }
}

function triggerPushMerge(node: Node, payload: Record<string, unknown>): TriggerResult {
  const p = (node.path || payload.path) as string;
  if (!p || !fs.existsSync(p)) return { ok: false, error: "missing path" };
  const remote = (payload.remote as string) || "origin";
  const branch = (payload.branch as string) || "main";
  const args = (payload.args as string[]) || [];
  try {
    execSync(`git -C ${JSON.stringify(p)} merge ${branch}`, { stdio: "inherit" });
    execSync(`git -C ${JSON.stringify(p)} push ${remote} ${args.join(" ")}`, { stdio: "inherit" });
    return { ok: true };
  } catch {
    return { ok: false };
  }
}

function triggerHpccJob(node: Node, payload: Record<string, unknown>): TriggerResult {
  const p = node.path || (payload.path as string);
  const script = (payload.script as string) || "job.sh";
  if (!p) return { ok: false, error: "missing path" };
  try {
    execSync(`sbatch ${path.join(p, script)}`, { cwd: p, stdio: "inherit" });
    return { ok: true };
  } catch {
    return { ok: false };
  }
}

function triggerRunScript(node: Node, payload: Record<string, unknown>): TriggerResult {
  const p = node.path || (payload.path as string);
  const script = payload.script as string;
  if (!p || !script) return { ok: false, error: "missing path/script" };
  try {
    execSync(script, { cwd: p, stdio: "inherit" });
    return { ok: true };
  } catch {
    return { ok: false };
  }
}
